### ISSUE TITLE
Failed to legalize `mhlo.select_and_scatter` op in gradient of pooling layer

### ISSUE THREAD

**Issue by phoenix-meadowlark** [iree-org/iree#10841]
### What happened?

IREE fails to compile training for models with pooling ops due to illegal `mhlo.select_and_scatter` ops. Support for this is needed replicate the DNN baseline in [Streaming keyword spotting on mobile devices](https://arxiv.org/pdf/2005.06720.pdf) with on-device training.

### Steps to reproduce your issue

The python to produce the MLIR below is as follows:

```python
import iree.jax
import jax
import jax.numpy as jnp
from praxis import layers

features = jnp.ones((2, 8, 1, 1))
pooling_p = layers.Pooling.HParams(
    name="pooling",
    window_shape=(2, 1),
    window_stride=(2, 1),
)
pooling_layer = pooling_p.Instantiate()


class PoolingProgram(iree.jax.Program):

  @iree.jax.kernel
  def _pooling_grad(features):
    def pool_and_sum(features):
      features, _ = pooling_layer.apply({}, features)
      return jnp.sum(features)
    return jax.value_and_grad(pool_and_sum)(features)

  def pooling_grad(self, features=iree.jax.like(features)):
    return self._pooling_grad(features)


program = PoolingProgram()
ir_module = iree.jax.Program.get_mlir_module(program)
with open("/tmp/pooling_grad.mlir", "wb") as f:
  ir_module.operation.print(file=f, binary=True)
```

```mlir
module @pooling {
  func.func @pooling_grad(%arg0: tensor<2x8x1x1xf32>) -> (tensor<f32>, tensor<2x8x1x1xf32>) {
    %0:2 = call @jit__pooling_grad$main(%arg0) : (tensor<2x8x1x1xf32>) -> (tensor<f32>, tensor<2x8x1x1xf32>)
    return %0#0, %0#1 : tensor<f32>, tensor<2x8x1x1xf32>
  }
  func.func private @jit__pooling_grad$main(%arg0: tensor<2x8x1x1xf32>) -> (tensor<f32>, tensor<2x8x1x1xf32>) {
    %0 = mhlo.constant dense<0xFF800000> : tensor<f32>
    %1 = "mhlo.broadcast"(%0) {broadcast_sizes = dense<> : tensor<0xi64>} : (tensor<f32>) -> tensor<f32>
    %2 = "mhlo.reduce_window"(%arg0, %1) ({
    ^bb0(%arg1: tensor<f32>, %arg2: tensor<f32>):
      %9 = mhlo.maximum %arg1, %arg2 : tensor<f32>
      mhlo.return %9 : tensor<f32>
    }) {
      base_dilations = dense<1> : tensor<4xi64>,
      padding = dense<0> : tensor<4x2xi64>,
      window_dilations = dense<1> : tensor<4xi64>,
      window_dimensions = dense<[1, 2, 1, 1]> : tensor<4xi64>,
      window_strides = dense<[1, 2, 1, 1]> : tensor<4xi64>
    } : (tensor<2x8x1x1xf32>, tensor<f32>) -> tensor<2x4x1x1xf32>
    %3 = mhlo.constant dense<0.000000e+00> : tensor<f32>
    %4 = mhlo.reduce(%2 init: %3) across dimensions = [0, 1, 2, 3] : (tensor<2x4x1x1xf32>, tensor<f32>) -> tensor<f32>
     reducer(%arg1: tensor<f32>, %arg2: tensor<f32>)  {
      %9 = mhlo.add %arg1, %arg2 : tensor<f32>
      mhlo.return %9 : tensor<f32>
    }
    %5 = mhlo.constant dense<1.000000e+00> : tensor<f32>
    %6 = "mhlo.broadcast_in_dim"(%5) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<f32>) -> tensor<2x4x1x1xf32>
    %7 = mhlo.constant dense<0.000000e+00> : tensor<f32>
    %8 = "mhlo.select_and_scatter"(%arg0, %6, %7) ({
    ^bb0(%arg1: tensor<f32>, %arg2: tensor<f32>):
      %9 = mhlo.compare  GE, %arg1, %arg2,  FLOAT : (tensor<f32>, tensor<f32>) -> tensor<i1>
      mhlo.return %9 : tensor<i1>
    }, {
    ^bb0(%arg1: tensor<f32>, %arg2: tensor<f32>):
      %9 = mhlo.add %arg1, %arg2 : tensor<f32>
      mhlo.return %9 : tensor<f32>
    }) {
      padding = dense<0> : tensor<4x2xi64>,
      window_dimensions = dense<[1, 2, 1, 1]> : tensor<4xi64>,
      window_strides = dense<[1, 2, 1, 1]> : tensor<4xi64>
    } : (tensor<2x8x1x1xf32>, tensor<2x4x1x1xf32>, tensor<f32>) -> tensor<2x8x1x1xf32>
    return %4, %8 : tensor<f32>, tensor<2x8x1x1xf32>
  }
}
```

`iree-compile` error (backend independent):

```mlir
/tmp/iree/hotword/pooling_grad.mlir:29:10: error: failed to legalize operation 'mhlo.select_and_scatter' that was explicitly marked illegal
    %8 = "mhlo.select_and_scatter"(%arg0, %6, %7) ({
         ^
/tmp/iree/hotword/pooling_grad.mlir:3:12: note: called from
    %0:2 = call @jit__pooling_grad$main(%arg0) : (tensor<2x8x1x1xf32>) -> (tensor<f32>, tensor<2x8x1x1xf32>)
           ^
/tmp/iree/hotword/pooling_grad.mlir:29:10: note: see current operation:
%17 = "mhlo.select_and_scatter"(%arg0, %1, %3) ({
^bb0(%arg1: tensor<f32>, %arg2: tensor<f32>):
  %18 = "mhlo.compare"(%arg1, %arg2) {compare_type = #mhlo<comparison_type FLOAT>, comparison_direction = #mhlo<comparison_direction GE>} : (tensor<f32>, tensor<f32>) -> tensor<i1>
  "mhlo.return"(%18) : (tensor<i1>) -> ()
}, {
^bb0(%arg1: tensor<f32>, %arg2: tensor<f32>):
  %18 = "mhlo.add"(%arg1, %arg2) : (tensor<f32>, tensor<f32>) -> tensor<f32>
  "mhlo.return"(%18) : (tensor<f32>) -> ()
}) {padding = dense<0> : tensor<4x2xi64>, window_dimensions = dense<[1, 2, 1, 1]> : tensor<4xi64>, window_strides = dense<[1, 2, 1, 1]> : tensor<4xi64>} : (tensor<2x8x1x1xf32>, tensor<2x4x1x1xf32>, tensor<f32>) -> tensor<2x8x1x1xf32>
    %8 = "mhlo.select_and_scatter"(%arg0, %6, %7) ({
         ^
compilation failed
```

### What component(s) does this issue relate to?

Frontends, MLIR, Compiler

### Version information

4d1b61368ce94b254b57caba66b0cca5b3bda032

**Comment by LiqinWeng** [comment]
Hello, is there any progress on this BUG fix? @rsuderman

**Comment by rsuderman** [comment]
There is a basic implementation for non-overlapping `mhlo.select_and_scatter` now implemented. Overlapping regions will require some more work but for most use cases the basic implementation should be fine.

**Comment by LiqinWeng** [comment]
> There is a basic implementation for non-overlapping `mhlo.select_and_scatter` now implemented. Overlapping regions will require some more work but for most use cases the basic implementation should be fine.

Hi，I also implemented a simple version, but I am lowering mhlo.slect_and_scatter to linalg_ext.select_and_scatter, Can you tell me what are the pros and cons of these two methods? @rsuderman

**Comment by LiqinWeng** [comment]
Here is my implemented a simple version: 
https://github.com/LiqinWeng/iree/commit/6946998d50dd40777117f2f3fb2de0286290b8d9
could you review it?@@rsuderman

**Comment by bviyer** [comment]
Here is a sample code that @rsuderman sent me to reproduce this issue:
```
import iree.jax
import jax
import flax
import numpy

matrix = numpy.random.random((1, 9, 9, 4)).astype(jax.numpy.single)


def test_function(matrix):

    def grad(matrix):
        return jax.numpy.sum(
            flax.linen.max_pool(matrix, (3, 2), (2, 2), 'VALID') -
            flax.linen.avg_pool(matrix, (3, 2), (2, 2), 'VAlID'))

    return jax.grad(grad)(matrix)


class Program(iree.jax.Program):

    @iree.jax.kernel
    def _test(x):
        return test_function(x)

    def test(self, x=iree.jax.like(matrix)):
        return self._test(x)


iree_program = Program()
iree_compiled = iree.jax.IREE.compile_program(iree_program)

ans_jax = jax.jit(test_function, backend="cpu")
ans_iree = iree_compiled.test

print(iree.jax.Program.get_mlir_module(iree_program))

print(ans_jax(matrix))
print(ans_iree(matrix).__array__())
```
Using this I was able to create the following mhlo file:

```
#loc = loc(unknown)
module {
  func.func @test(%arg0: tensor<1x9x9x4xf32> loc(unknown)) -> tensor<1x9x9x4xf32> {
    %0 = call @jit__test$main(%arg0) : (tensor<1x9x9x4xf32>) -> tensor<1x9x9x4xf32> loc(#loc)
    return %0 : tensor<1x9x9x4xf32> loc(#loc)
  } loc(#loc)
  func.func private @jit__test$main(%arg0: tensor<1x9x9x4xf32> {jax.arg_info = "x", mhlo.sharding = "{replicated}"} loc(unknown)) -> (tensor<1x9x9x4xf32> {jax.result_info = "", mhlo.sharding = "{replicated}"}) {
    %0 = stablehlo.constant dense<6> : tensor<i32> loc(#loc)
    %1 = stablehlo.convert %0 : (tensor<i32>) -> tensor<f32> loc(#loc5)
    %2 = stablehlo.constant dense<1.000000e+00> : tensor<f32> loc(#loc)
    %3 = stablehlo.broadcast_in_dim %2, dims = [] : (tensor<f32>) -> tensor<1x4x4x4xf32> loc(#loc6)
    %4 = stablehlo.negate %3 : tensor<1x4x4x4xf32> loc(#loc7)
    %5 = stablehlo.broadcast_in_dim %1, dims = [] : (tensor<f32>) -> tensor<1x4x4x4xf32> loc(#loc8)
    %6 = stablehlo.divide %4, %5 : tensor<1x4x4x4xf32> loc(#loc8)
    %7 = stablehlo.constant dense<0.000000e+00> : tensor<f32> loc(#loc)
    %8 = stablehlo.pad %6, %7, low = [0, 2, 1, 0], high = [0, 2, 2, 0], interior = [0, 1, 1, 0] : (tensor<1x4x4x4xf32>, tensor<f32>) -> tensor<1x11x10x4xf32> loc(#loc9)
    %9 = stablehlo.constant dense<0.000000e+00> : tensor<f32> loc(#loc10)
    %10 = stablehlo.broadcast_in_dim %9, dims = [] : (tensor<f32>) -> tensor<f32> loc(#loc10)
    %11 = "stablehlo.reduce_window"(%8, %10) ({
    ^bb0(%arg1: tensor<f32> loc(unknown), %arg2: tensor<f32> loc(unknown)):
      %15 = stablehlo.add %arg1, %arg2 : tensor<f32> loc(#loc10)
      stablehlo.return %15 : tensor<f32> loc(#loc10)
    }) {base_dilations = dense<1> : tensor<4xi64>, padding = dense<0> : tensor<4x2xi64>, window_dilations = dense<1> : tensor<4xi64>, window_dimensions = dense<[1, 3, 2, 1]> : tensor<4xi64>, window_strides = dense<1> : tensor<4xi64>} : (tensor<1x11x10x4xf32>, tensor<f32>) -> tensor<1x9x9x4xf32> loc(#loc10)
    %12 = stablehlo.constant dense<0.000000e+00> : tensor<f32> loc(#loc11)
    %13 = "stablehlo.select_and_scatter"(%arg0, %3, %12) ({
    ^bb0(%arg1: tensor<f32> loc(unknown), %arg2: tensor<f32> loc(unknown)):
      %15 = stablehlo.compare  GE, %arg1, %arg2,  FLOAT : (tensor<f32>, tensor<f32>) -> tensor<i1> loc(#loc13)
      stablehlo.return %15 : tensor<i1> loc(#loc12)
    }, {
    ^bb0(%arg1: tensor<f32> loc(unknown), %arg2: tensor<f32> loc(unknown)):
      %15 = stablehlo.add %arg1, %arg2 : tensor<f32> loc(#loc14)
      stablehlo.return %15 : tensor<f32> loc(#loc12)
    }) {padding = dense<0> : tensor<4x2xi64>, window_dimensions = dense<[1, 3, 2, 1]> : tensor<4xi64>, window_strides = dense<[1, 2, 2, 1]> : tensor<4xi64>} : (tensor<1x9x9x4xf32>, tensor<1x4x4x4xf32>, tensor<f32>) -> tensor<1x9x9x4xf32> loc(#loc12)
    %14 = stablehlo.add %11, %13 : tensor<1x9x9x4xf32> loc(#loc15)
    return %14 : tensor<1x9x9x4xf32> loc(#loc)
  } loc(#loc)
} loc(#loc)
#loc1 = loc("/usr/local/google/home/bviyer/LLVM/my_env/lib/python3.10/site-packages/flax/linen/pooling.py":93:0)
#loc2 = loc("/usr/local/google/home/bviyer/LLVM/My_Fork/iree/build/pooling_grad_small.py":12:0)
#loc3 = loc("/usr/local/google/home/bviyer/LLVM/My_Fork/iree/build/pooling_grad_small.py":13:0)
#loc4 = loc("/usr/local/google/home/bviyer/LLVM/my_env/lib/python3.10/site-packages/flax/linen/pooling.py":69:0)
#loc5 = loc("jit(_test)/jit(main)/convert_element_type[new_dtype=float32 weak_type=False]"(#loc1))
#loc6 = loc("jit(_test)/jit(main)/broadcast_in_dim[shape=(1, 4, 4, 4) broadcast_dimensions=()]"(#loc2))
#loc7 = loc("jit(_test)/jit(main)/neg"(#loc3))
#loc8 = loc("jit(_test)/jit(main)/div"(#loc1))
#loc9 = loc("jit(_test)/jit(main)/pad[padding_config=((0, 0, 0), (2, 2, 1), (1, 2, 1), (0, 0, 0))]"(#loc4))
#loc10 = loc("jit(_test)/jit(main)/reduce_window_sum[window_dimensions=(1, 3, 2, 1) window_strides=(1, 1, 1, 1) padding=((0, 0), (0, 0), (0, 0), (0, 0)) base_dilation=(1, 1, 1, 1) window_dilation=(1, 1, 1, 1)]"(#loc4))
#loc11 = loc("jit(_test)/jit(main)/select_and_scatter_add[select_prim=ge window_dimensions=(1, 3, 2, 1) window_strides=(1, 2, 2, 1) padding=((0, 0), (0, 0), (0, 0), (0, 0))]"(#loc4))
#loc12 = loc("jit(_test)/jit(main)/select_and_scatter[select_consts=() scatter_consts=() window_dimensions=(1, 3, 2, 1) window_strides=(1, 2, 2, 1) padding=((0, 0), (0, 0), (0, 0), (0, 0))]"(#loc4))
#loc13 = loc("jit(_test)/jit(main)/ge"(#loc4))
#loc14 = loc("jit(_test)/jit(main)/add"(#loc4))
#loc15 = loc("jit(_test)/jit(main)/add_any"(#loc4))
```

When I compiled it using following flags, I was able to reproduce the error. Here is my screenshot:
```
$ iree-compile --iree-input-type=mhlo /tmp/core-reproducer.mlir -o /tmp/out.vmfb
/usr/local/google/home/bviyer/LLVM/my_env/lib/python3.10/site-packages/flax/linen/pooling.py:69:0: error: failed to legalize operation 'mhlo.select_and_scatter' that was explicitly marked illegal
/usr/local/google/home/bviyer/LLVM/my_env/lib/python3.10/site-packages/flax/linen/pooling.py:69:0: note: see current operation: 
%12 = "mhlo.select_and_scatter"(%arg0, %3, %5) ({
^bb0(%arg1: tensor<f32>, %arg2: tensor<f32>):
  %14 = "mhlo.compare"(%arg1, %arg2) {compare_type = #mhlo<comparison_type FLOAT>, comparison_direction = #mhlo<comparison_direction GE>} : (tensor<f32>, tensor<f32>) -> tensor<i1>
  "mhlo.return"(%14) : (tensor<i1>) -> ()
}, {
^bb0(%arg1: tensor<f32>, %arg2: tensor<f32>):
  %14 = "mhlo.add"(%arg1, %arg2) : (tensor<f32>, tensor<f32>) -> tensor<f32>
  "mhlo.return"(%14) : (tensor<f32>) -> ()
}) {padding = dense<0> : tensor<4x2xi64>, window_dimensions = dense<[1, 3, 2, 1]> : tensor<4xi64>, window_strides = dense<[1, 2, 2, 1]> : tensor<4xi64>} : (tensor<1x9x9x4xf32>, tensor<1x4x4x4xf32>, tensor<f32>) -> tensor<1x9x9x4xf32>
$
```

**Comment by bviyer** [comment]
Here's a further shorter test case:
```
func.func @test(%3: tensor<1x4x4x4xf32>, %arg0: tensor<1x9x9x4xf32>) -> tensor<1x9x9x4xf32> {
    %12 = stablehlo.constant dense<0.000000e+00> : tensor<f32>
    %13 = "stablehlo.select_and_scatter"(%arg0, %3, %12) ({
    ^bb0(%arg1: tensor<f32>, %arg2: tensor<f32>):
      %15 = stablehlo.compare  GE, %arg1, %arg2,  FLOAT : (tensor<f32>, tensor<f32>) -> tensor<i1>
      stablehlo.return %15 : tensor<i1>
    }, {
    ^bb0(%arg1: tensor<f32> , %arg2: tensor<f32> ):
      %15 = stablehlo.add %arg1, %arg2 : tensor<f32>
      stablehlo.return %15 : tensor<f32>
    }) {padding = dense<0> : tensor<4x2xi64>, window_dimensions = dense<[1, 3, 2, 1]> : tensor<4xi64>, window_strides = dense<[1, 2, 2, 1]> : tensor<4xi64>} : (tensor<1x9x9x4xf32>, tensor<1x4x4x4xf32>, tensor<f32>) -> tensor<1x9x9x4xf32>
    return %13 : tensor<1x9x9x4xf32>
}
```

**Comment by bviyer** [comment]
I looked at this work a bit, I think its best to do a stablehlo to stable hlo conversion instead of doing it in linalg. The overall idea is to break down the large array into the tile of source based on the size of the window and call the select function on this and then do scatter.

**Comment by benvanik** [comment]
You'll want to pay careful attention to fusion - selects that don't get fused with both producers are likely to materialize both options in global memory before doing the select. That is to say, a loose graph of stablehlo ops with selects in it is a potential perf/memory consumption footgun.

**Comment by debjyoti0891** [comment]
Is there an update on this? I recently encountered the same bug with iree-compile version `3.2.0rc20250206`
